File: shortest-path/cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        // dp[i][j]: cheapest cost to dst j within i+1 flight.
        vector<vector<int>> dp(k+2, vector<int>(n, INT_MAX/2));
        dp[0][src] = 0;
        int res = INT_MAX/2;
        for (int i=1; i<=k+1; i++){
            for (auto flight:flights){
                int a = flight[0], b = flight[1], p = flight[2];
                dp[i][b] = min(dp[i][b], dp[i - 1][a] + p);
                if (b==dst) res = min(res, dp[i][b]);
            }
        }
        return res == INT_MAX/2 ? -1:res;
    }
};

File: heap-priority-queue/cheapest-flights-within-k-stops.py
from collections import defaultdict

class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        graph = defaultdict(list)
        for start, end, price in flights:
            graph[start].append((end, price))
       
        # dist[m][n] = min cost from src to m when used n steps
        dist = [[float('inf')] * (k + 1) for _ in range(n)]
        dist[src][0] = 0
        # heap = [price, curr_stop, step_used]
        heap = [(0, src, 0)]
        while heap:
            price, stop, step = heapq.heappop(heap) 

            if stop == dst:
                return price
            if price > dist[stop][step]:
                continue
            if step <= k:
                for neighbor, p in graph[stop]:
                    heapq.heappush(heap, (price + p, neighbor, step + 1))
        return -1

